Tests/atlantis

meowsbits opened the pull request:
This is an experimental changeset that at least templates what's needed to run tests beyond those included in the ETH Foundation-specific suite ethereum/tests (which is as a submodule in the repo as `tests/testData`). 

- Move submodule remote origin to multi-geth/tests. I've been using multi-geth to fill tests, and have pushed changes there. Relevant changes are currently held there at feature branch `feat/etc-atlantis`; the submodule's HEAD is moved to that as well. 
- Adds hardcoded test chain configs. Ideally we can in the future move to reading chain specs on a per test, or per test suite, basis, removing the need for these "you just have to know how to configure it" lookups and technical debt. The tests submodule here also includes an example `chainspecs/` dir that intends to make a suggestion toward this. 

And... the tests pass, btw :) This means that we're able to use existing Byzantium test suites to create Atlantis tests (b/c the tests generator is hacky and piggy-backs existing state pre, envs, and transitions), and use multi-geth to create test target values that classic geth agrees on. With a few modifications to Parity's test runner, we should be able to extend this idea there as well.

meowsbits committed:
tests: update tests submodule to latest 'develop' branch head

meowsbits committed:
.gitmodules: swap tests remote from ethereum -> multi-geth

meowsbits committed:
tests: add and update tests to use experimental atlantis-configured tests
For state and difficulty test suites.
The testData submodule uses an experimental branch feat/etc-atlantis

soc1c commented on .gitmodules:
```diff
@@ -1,3 +1,3 @@
 [submodule "tests/testData"]
 	path = tests/testData
-	url = https://github.com/ethereum/tests.git
+	url = https://github.com/multi-geth/tests.git
```
:+1:

soc1c commented on tests/util.go:
```diff
@@ -112,7 +112,11 @@ func insertAccount(state *state.StateDB, saddr string, account Account) {
+		if account.Nonce == "" {
+			i = 0
+		} else {
+			panic(err)
```
what the case this would happen?

meowsbits replied:
afaiu, i think it's just a small issue with the way multi-geth encodes the json test output (0 value left empty instead of writing explicit 0 kind of thing)

GregTheGreek replied:
yeah im not sure about this?

meowsbits replied:
yea, ethereum/tests includes zero values:

https://github.com/ethereum/tests/blob/develop/GeneralStateTests/stArgsZeroOneBalance/addNonConst.json#L80-L93

but multi-geth writing tests in omitting them

https://github.com/multi-geth/tests/blob/feat/etc-atlantis/GeneralStateTests/stArgsZeroOneBalance/addNonConst.json#L17-L25

meowsbits replied:
and multi-geth's omissions are b/c of the way it encodes genesis account types:

```

// GenesisAccount is an account in the state of the genesis block.
type GenesisAccount struct {
	Code       []byte                      `json:"code,omitempty"`
	Storage    map[common.Hash]common.Hash `json:"storage,omitempty"`
	Balance    *big.Int                    `json:"balance" gencodec:"required"`
	Nonce      uint64                      `json:"nonce,omitempty"`
	PrivateKey []byte                      `json:"secretKey,omitempty"` // for tests
}
```
> https://github.com/multi-geth/multi-geth/blob/master/core/genesis.go#L80-L87

austinabell replied:
If there is no way to fill the test with nonces included I could be fine with this, I just would rather not do this for reasons of future proofing for possible tests in the future, as someone may not notice that the nonce is being defaulted to 0x0

meowsbits replied:
yea agreed. there's definitely a way to do it, it's just not a priority for this level of experiment so far. it is imaginable that multi-geth might get a cli command to generate these tests and then would be expected to match syntax exactly

meowsbits committed:
tests: fix to make state explicity set nonce

This was actually a noop before since 'i' value went unused and state
was left to implicit default behavior.

This changes is a noop, but makes the code right.

austinabell commented:
Awesome, this will be great to add to check compliance between clients.

One thing to note is there is still the discrepancy between clients for the EIP 684 (Not included in Ethereum classic) and the edge case of state trie clearing that the tests are currently being skipped. 

```go
	// EIP 684 Implementations
	skipTests["TransactionCollisionToEmptyButCode.json"] = "Not Implemented"
	skipTests["TransactionCollisionToEmpty.json"] = "Not Implemented"
	skipTests["TransactionCollisionToEmptyButNonce.json"] = "Not Implemented"
	skipTests["CreateCollisionToEmpty.json"] = "Not Implemented"
	skipTests["CreateHashCollision.json"] = "Not Implemented"
	skipTests["createJS_ExampleContract.json"] = "Not Implemented"
	skipTests["RevertDepthCreateAddressCollision.json"] = "Not Implemented"
```

and 

```go
	skipTests["RevertPrefoundEmptyOOG.json"] = "State trie clearing unimplemented"
	skipTests["FailedCreateRevertsDeletion.json"] = "State trie clearing unimplemented"
```

It was great to note from this that the issue regarding call operations touching accounts to be deleted (specifically STATICCALL) are compliant across clients as of now, although maybe not ideal if following the spec correctly this time and not handling faulty behaviour to achieve consensus #41

soc1c commented:
What's the status of this?
